src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,1519 @@
+				$this->parse_error();
+				dbg("Skipping AAA: formatting element is not in the stack of open elements", 2);
+
+				/**
+				 * This is not in the spec, but it's necessary.
+				 * 
+				 * If we were building a DOM, moving on without 
+				 * creating a Node would be the same as dropping 
+				 * the unexpected token.
+				 * 
+				 * We're processing a text stream, though, so simply
+				 * moving on would leave that token in place. Instead,
+				 * we need to drop it explicitly.
+				 */
+				$this->drop_current_tag_token();
+				return;
+			}
+
+			// If formatting element is not in scope, then this is a parse error; return
+			if ( ! $this->is_element_in_scope( $formatting_element ) ) {
+				$this->parse_error();
+				dbg("Skipping AAA: formatting element {$formatting_element->tag} is not in scope", 2);
+				
+				/**
+				 * This is not in the spec, but it's necessary.
+				 * See the previous "if" statement for details.
+				 */
+				$this->drop_current_tag_token();
+				return;
+			}
+
+			// If formatting element is not the current node, then this is a parse error.
+			// (But do not return.)
+			if ( $formatting_element !== $this->current_node() ) {
+				$this->parse_error();
+			}
+
+			/*
+			 * Let furthest block be the topmost node in the stack of open elements that
+			 * is lower in the stack than formatting element, and is an element in the
+			 * special category. There might not be one.
+			 */
+			$furthest_block = null;
+			for ( $i = count( $this->open_elements ) - 1; $i >= 0; $i-- ) {
+				$node = $this->open_elements[ $i ];
+				if ( $node === $formatting_element ) {
+					break;
+				}
+				if ( $this->is_special_element( $node->tag ) ) {
+					$furthest_block = $node;
+				}
+			}
+
+			// If there is no such node, then the UA must first pop all the nodes from
+			// the bottom of the stack of open elements, from the current node up to
+			// and including formatting element, then remove formatting element from
+			// the list of active formatting elements, and finally abort these steps.
+			if ( null === $furthest_block ) {
+				$this->pop_until_node( $formatting_element, false );
+				array_splice( $this->active_formatting_elements, $formatting_element_idx, 1 );
+				dbg("Skipping AAA: no furthest block found", 2);
+				return;
+			}
+
+			// We didn't bale out so far, but the algorithm is not implemented.
+			// Let's error out.
+			break;
+		}
+		throw new Exception('Adoption Agency Algorithm not supported.');
+	}
+
+	private function insert_element( WP_HTML_Tag_Token $token ) {
+		if($token !== $this->current_token) {
+			// Aesthetic choice for now.
+			// @TODO: discuss it with the team
+			$tag = strtolower($token->tag);
+
+			$this->add_lexical_update(
+				new WP_HTML_Text_Replacement(
+					$this->current_token_start,
+					$this->current_token_start,
+					"<{$tag}>"
+				)
+			);
+		}
+		array_push($this->open_elements, $token);
```

Comment:
I think this doesn't matter right now, but I'm strongly opposed to offering a function which would allow people to add a partial tag, an opener without a closer. adding a void or self-closing foreign element seems fine, but leaving an open element on the stack seems to be an invitation for trouble

Reply:
I agree! I didn't mean to offer this method to people. `insert_element` is a private API that implements [insert an HTML element for a token](https://html.spec.whatwg.org/multipage/parsing.html#insert-an-html-element) part of the parsing spec.